### PR TITLE
Add redis cluster cache handling

### PR DESCRIPTION
**Pull request by hufon**
- For session, cache and ratelimiter.

Using Redis in cluster mode (not sentinel mode) requires to use Redis.Cluster from ioredis.
It's enabled by setting `CACHE_REDIS_CLUSTER`/`SESSION_REDIS_CLUSTER`/`RATE_LIMITER_CLUSTER` to true as env variables

### CODE REVIEW

**Comment by rijkvanzanten** [comment]
> For session, cache and ratelimiter.

Can we make it individual env vars for these? You can theoretically have multiple redis instances for the individual parts, and not use clusters for all of them

**Comment by hufon** [comment]
> > For session, cache and ratelimiter.
> 
> Can we make it individual env vars for these? You can theoretically have multiple redis instances for the individual parts, and not use clusters for all of them

what do you suggest for naming?

**Comment by rijkvanzanten** [comment]
> (as i can't use XXX_REDIS_XXX variables)

That's exactly what I was going to suggest. Using `CACHE_REDIS_CLUSTER` etc. The `getConfigFromEnv` function allows you to ignore specific env vars, so we should be able to use that

**Comment by hufon** [comment]
> > (as i can't use XXX_REDIS_XXX variables)
> 
> That's exactly what I was going to suggest. Using `CACHE_REDIS_CLUSTER` etc. The `getConfigFromEnv` function allows you to ignore specific env vars, so we should be able to use that

done

File: api/src/cache.ts [hunk 1]
```diff
@@ -52,11 +53,17 @@ function getConfig(
...
-		});
+		const KeyvAnyRedis = require('keyv-anyredis');
+		const Redis = require('ioredis');
+		if (env.CACHE_REDIS_CLUSTER === true) {
+			const client = new Redis.Cluster(env.CACHE_REDIS || getConfigFromEnv('CACHE_REDIS_', 'CACHE_REDIS_CLUSTER'), {
+				dnsLookup: (address: any, callback: (arg0: null, arg1: any) => any) => callback(null, address),
+			});
+			config.store = new KeyvAnyRedis(client);
+		} else {
+			const client = new Redis(env.CACHE_REDIS || assign(getConfigFromEnv('CACHE_REDIS_'), { commandTimeout: 500 }));
```
Comment (azrikahar) [inline api/src/cache.ts:64]: Note: our existing KeyvRedis had the option `{ commandTimeout: 500 }` and it'll use `Object.assign()` under the hood here: https://github.com/lukechilds/keyv-redis/blob/9d650f21ce4609c07d5e6848efb1572eccf95d46/src/index.js#L13

So this lodash `assign` was used to keep this option intact & resolve merge conflict.

**Comment by aidenfoxx** [comment]
As far as I know "@keyv/redis" supports Redis.Cluster? Any reason we're changing the package? https://github.com/jaredwray/keyv/blob/master/packages/redis/src/index.js

**Comment by azrikahar** [comment]
> As far as I know "@keyv/redis" supports Redis.Cluster? Any reason we're changing the package? https://github.com/jaredwray/keyv/blob/master/packages/redis/src/index.js

Interesting. Admittedly this PR has been outstanding for a while now, but from my surface level check, seems like technically that `uri instanceof Redis.Cluster` line specifically was added by end of last year which is months after this PR. But you are right in that we should reconsider this point when we can merge this. Once again apologies for the delay on this.

**Comment by jesusmpc** [comment]
Any info on when Redis Cluster support will be included?

**Comment by rijkvanzanten** [comment]
Heya, thanks for your PR! Our small team has been attempting to work through a backlog of external contributions for the better part of a year, but in that time many have become too stale to merge or the change is no longer desired/relevant.. 
We've updated the [contributing guidelines](https://github.com/directus/directus/blob/main/contributing.md#pull-requests) and are working on [making more resources available](https://github.com/directus/directus/discussions/17977) to make sure that doesn’t happen again 😬 In the meantime, we'll have to close this PR for now..

**Pull request closed without merging**